azkaban-web-server/src/test/java/azkaban/webapp/servlet/ExecutorServletTest.java

```diff
@@ -0,0 +1,125 @@
+
+  private ExecutorServlet executorServlet;
+
+  @Override
+  @Before
+  public void setUp() throws Exception {
+    super.setUp();
+    this.executorServlet = new ExecutorServlet();
+    this.executorServlet.init(this.servletConfig);
+  }
+
+  /**
+   * When executeFlow is called with SLA options, a new execution should be submitted with SLA.
+   */
+  @Test
+  public void postAjaxExecuteFlowWithSlaSettings() throws Exception {
+    mockTestProjectAndFlow();
+    mockSubmitExecution();
+    this.req.addParameter("ajax", "executeFlow");
+    this.req.addParameter("project", "testProject");
+    this.req.addParameter("flow", "testFlow");
+    this.req.addParameter("slaSettings[1]", ",FINISH,2:30,true,false");
+    this.req.addParameter("slaEmails", "sla1@example.com,sla2@example.com");
+
+    this.executorServlet.handlePost(this.req, this.res, this.session);
+
+    assertEquals(200, this.res.getStatus());
+
+    final JsonNode json = this.res.getResponseJson();
+    assertEquals("testProject", json.path("project").asText());
+    assertEquals("testFlow", json.path("flow").asText());
+    assertEquals("Submitted (mocked)", json.path("message").asText());
+    assertEquals(99, json.path("execid").asInt());
+
+    final List<SlaOption> slaOptions = this.exFlow.getValue().getExecutionOptions().getSlaOptions();
+    final List<SlaOption> expected = Arrays.asList(new SlaOption(SlaType.FLOW_FINISH, "testFlow",
+        "", Duration.ofMinutes(150), ImmutableSet.of(SlaAction.ALERT),
+        ImmutableList.of("sla1@example.com", "sla2@example.com")));
+    Assert.assertEquals(expected, slaOptions);
+  }
+
+  /**
+   * When SLA options can't be parsed, API is expected to return 200 OK with JSON that has "error".
+   */
+  @Test
+  public void postAjaxExecuteFlowWithInvalidSlaSettings() throws Exception {
+    mockTestProjectAndFlow();
+    mockSubmitExecution();
+    this.req.addParameter("ajax", "executeFlow");
+    this.req.addParameter("project", "testProject");
+    this.req.addParameter("flow", "testFlow");
+    this.req.addParameter("slaSettings[1]", "broken-syntax-for-sla");
+
+    this.executorServlet.handlePost(this.req, this.res, this.session);
+
+    assertEquals(200, this.res.getStatus());
+    assertEquals("Error parsing flow options: Error parsing SLA setting 'broken-syntax-for-sla': "
+            + "java.lang.ArrayIndexOutOfBoundsException: 1",
+        this.res.getResponseJson().path("error").asText());
+  }
+
+  /**
+   * When executeFlow is called without project param, API is expected to raise an exception.
+   */
+  @Test
+  public void postAjaxExecuteFlow() throws Exception {
+    this.req.addParameter("ajax", "executeFlow");
+    try {
+      this.executorServlet.handlePost(this.req, this.res, this.session);
+      fail("Expected exception was not thrown");
+    } catch (final ServletException e) {
+      assertEquals("Missing required parameter 'project'.", e.getMessage());
```

Comment:
@ypadron-in I added this unit test for ExecutorServlet. Something I noticed is that `executeFlow` raises an exception in some cases. Would the right thing to do even in this case to return 200 OK with an "error" JSON? Any way, I'm _not_ proposing to change any of that existing behavior as part of this PR.

Reply:
Thanks for adding the tests! 
You're right, 200 status code shouldn't be returned when there is an error but setting the right status code means adding a breaking change for the API and unfortunately there isn't a versioning strategy yet.
Returning an exception instead of a json response is also incorrect :(.